Add a macro `pg_cast` for constructing PG casts from Rust functions.

xwkuang5 opened the pull request:
To use the new macro, add the `pg_cast` macro on a single-argument single-return value Rust function:

```rust
#[pg_cast]
fn test_cast(_value: Json) -> i32 {
    0
}
```

It is possible to modify the cast by adding either `assignment` or `implicit` to the `pg_cast` macro as argument:

```
#[pg_cast(assignment)] /* creates a PG cast that may implicitly be used in assignment */

#[pg_cast(implicit)] /* creates a PG cast that may implicitly be used in any situation */
```

TESTED="cargo test"

xwkuang5 committed:
Add a macro `pg_cast` for constructing PG casts from Rust functions.

To use the new macro, add the `pg_cast` and `pg_extern` macros on
single-argument single-return value Rust function:

```rust
[pg_extern]
[pg_cast]
fn test_cast(_value: Json) -> i32 {
    0
}
```

It is possible to parametrize the cast by adding either `assignment` or
`implicit` to the `pg_cast` macro as argument:

```
[pg_cast(assignment)] -> creates a PG cast that can be used for casting
from one type to another at assignment time

[pg_cast(implicit)] -> creates a PG cast that can be used for casting
from one type to another in any context
```

TESTED=`cargo-pgrx pgrx test --features "pg13"`

eeeebbbbrrrr commented:
Thanks. Best Christmas present ever! 

I won’t have a chance to review it probably until January, so hang in there.

workingjubilee commented:
why does `pg_cast` not emit `pg_extern` if that is required?

xwkuang5 commented:
> why does `pg_cast` not emit `pg_extern` if that is required?

I considered 3 implementation options:

1. (current): use `pg_extern` + `pg_cast` as a macro for using the generated function in a PG cast
2. make `pg_cast` emit `pg_extern` and add another `castoption` macro for modifying the cast object
3. make `pg_cast` emit `pg_extern` and add `Implicit` and `Assignment` as an option to [Attribute](https://github.com/pgcentralfoundation/pgrx/blob/614456e191a87f6df1b72428c79a01dedfe75516/pgrx-sql-entity-graph/src/pg_extern/attribute.rs#L27)

Option 2 requires 2 more macros which seems unnecessary to me (though this is the approach that `pg_operator` takes). For option 3, I was not sure about whether adding the generic `Implicit` and `Assignment` variant to the `Attribute` enum is a good idea as I'm not sure if there will be any other PG objects that share the same attributes. I decided on option 1 because I felt that `pg_extern` is already exposed as a API that users know how to use. So `pg_cast` can focus on just adding the `CREATE CAST` statement. Currently, using `pg_cast` without `pg_extern` or using `pg_extern` without `pg_cast` both do nothing so the "error" handling seems okay to me.

I may be missing something or not grasping the overall macro design consideration here so curious about what you think!

workingjubilee commented:
Option 2 does seem bad, yes.

> Currently, using `pg_cast` without `pg_extern` or using `pg_extern` without `pg_cast` both do nothing so the "error" handling seems okay to me.

Oh, we wouldn't want to accept using `#[pg_cast]` without `#[pg_extern]`. In that case, because the user specified an intent to designate a cast, but hasn't designated a valid target of a cast, we should error.

Obviously, that's a one-way constraint. The other way around wouldn't suddenly emit errors, since that would break... uh, every instance of `#[pg_extern]` ever.

In general, people don't actually _write_ the `#[pg_operator]` invocations. It's available, but for the most part they are instead part of the special `derive`s, so it is tolerable if their API is confusing. That might help to keep in mind when interpreting our existing design.

For Option 3, it's fine if we make any number of breaking changes to that enum, so it remains a possibility.

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:
```diff
@@ -472,6 +474,122 @@ impl ToSql for PgExternEntity {
+                                                    line = self.line,
+                                                    name = self.name,
+                                                    module_path = self.module_path,
+                                                    source_arg_sql = source_arg_sql,
```
I prefer to avoid redundantly specifying names in formatting.
```suggestion
```

xwkuang5 replied:
Done

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:
```diff
@@ -472,6 +474,122 @@ impl ToSql for PgExternEntity {
+                                                    source_arg_sql = source_arg_sql,
+                                                    schema_prefix_source = context.schema_prefix_for(&source_arg_graph_index),
+                                                    source_name = source_arg.type_name,
+                                                    target_arg_sql = target_arg_sql,
```
```suggestion
```

xwkuang5 replied:
Done

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:
```diff
@@ -472,6 +474,122 @@ impl ToSql for PgExternEntity {
+                                                    schema_prefix_target = context.schema_prefix_for(&target_arg_graph_index),
+                                                    target_name = target_arg.type_name,
+                                                    function_name = self.name,
+                                                    optional = optional,
```
```suggestion
```

xwkuang5 replied:
Done

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/mod.rs:
```diff
@@ -230,6 +236,36 @@ impl PgExtern {
+                                match p.path.segments.last().unwrap().ident.to_string().as_str() {
+                                    "implicit" => cast = PgCast::Implicit,
+                                    "assignment" => cast = PgCast::Assignment,
+                                    _ => eprintln!("Unrecognized option: {}. Using default cast options.", p.path.to_token_stream()),
+                                }
+                            }
+                            _ => eprintln!(
+                                "Unable to parse attribute to pg_cast as a Rust Expr: {}. Using default cast options.",
+                                attr.tokens
+                            ),
```
I would prefer to not silently accept something that can be validated at compile time. Please return an error or panic instead.

xwkuang5 replied:
Done. Switched to panic on invalid options

workingjubilee reviewed (dismissed):
For the errors I am asking for, please validate they occur using `pgrx-tests/tests/ui` tests.

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:
```diff
@@ -472,6 +474,122 @@ impl ToSql for PgExternEntity {
+                    "PG cast function must have exactly one argument, got {}",
+                    self.fn_args.len()
+                ));
+            }
+            let source_arg = self
+                .metadata
+                .arguments
+                .first()
+                .ok_or_else(|| eyre!("Did not find source type for cast `{}`.", self.name))?;
+            let source_fn_arg = self
+                .fn_args
+                .first()
+                .ok_or_else(|| eyre!("Did not find source type for cast `{}`.", self.name))?;
+            let source_arg_graph_index = context
+                .graph
+                .neighbors_undirected(self_index)
+                .find(|neighbor| match &context.graph[*neighbor] {
+                    SqlGraphEntity::Type(ty) => ty.id_matches(&source_fn_arg.used_ty.ty_id),
+                    SqlGraphEntity::Enum(en) => en.id_matches(&source_fn_arg.used_ty.ty_id),
+                    SqlGraphEntity::BuiltinType(defined) => defined == source_arg.type_name,
+                    _ => false,
+                })
+                .ok_or_else(|| {
+                    eyre!("Could not find source type in graph. Got: {:?}", source_arg)
+                })?;
+            let source_arg_sql = match source_arg.argument_sql {
+                Ok(SqlMapping::As(ref sql)) => sql.clone(),
+                Ok(SqlMapping::Composite { array_brackets }) => {
+                    if array_brackets {
+                        let composite_type = self.fn_args[0].used_ty.composite_type
+                            .ok_or(eyre!("Found a composite type but macro expansion time did not reveal a name, use `pgrx::composite_type!()`"))?;
+                        format!("{composite_type}[]")
+                    } else {
+                        self.fn_args[0].used_ty.composite_type
+                            .ok_or(eyre!("Found a composite type but macro expansion time did not reveal a name, use `pgrx::composite_type!()`"))?.to_string()
+                    }
+                }
+                Ok(SqlMapping::Skip) => {
+                    return Err(eyre!("Found an skipped SQL type in a cast, this is not valid"))
+                }
```
At least one of these errors should also have a UI test for it. It's fine if you can't contrive a test case for each of these errors, I just want to see one.

xwkuang5 replied:
I added two basic tests in the `ui` directory but could not figure out how to run them. Could you share a command that I can use to invoke those UI tests? I also tried `cargo test --verbose --features "pg13" --no-default-features -- --list` to list the tests in `pgrx-tests` but could not find any ui test

workingjubilee commented on pgrx-macros/src/lib.rs:
```diff
@@ -148,6 +148,13 @@ pub fn initialize(_attr: TokenStream, item: TokenStream) -> TokenStream {
     item
 }
 
+/// Declare a function as `#[pg_cast]` to indicate that it represents a Postgres cast
+/// `cargo pgrx schema` will automatically generate the underlying SQL
+#[proc_macro_attribute]
+pub fn pg_cast(_attr: TokenStream, item: TokenStream) -> TokenStream {
+    item
+}
```
This should, if we go with `pg_cast` requiring `pg_extern`, specify that it requires `pg_extern`.

In either case, it should also document the valid arguments to the attribute (implicit, assignment), because these things are not made obvious by tooling, unlike how they would be in e.g. a function (which would emit its arguments and types into rustdoc, at least).

Not all `proc_macro_attribute`s in `pgrx-macros` do this, so you may be short on examples to imitate. It just has to cover the very basics, here.

xwkuang5 replied:
Went with a different approach of having `pg_cast` do the compile time validation and only after validation delegate to `PgExtern` for code-gen for the actual pg functions and casts. This avoids all the impossible states that the previous approach has.

Let me know what you think!

xwkuang5 committed:
`pg_cast` validate attributes at macro expansion time.

- Validate attributes at compile time before delegating to `PgExtern`
- Using `pg_cast` and `pg_extern` on the same item will result in
  duplicate definitions
- Added a `as_cast(PgCast)` option to `PgExtern` to augment `PgExtern`.
  Did not implement this as `PgExtern` attribute to avoid poluting
  `PgExtern` options

workingjubilee commented:
Nice! This seems pretty good. Going to take a closer look tomorrow, but it seems mergeable as-is or with perhaps small changes. I'm [considering revising the idiom for these attribute macros to match the SQL they implement](https://github.com/pgcentralfoundation/pgrx/issues/1488) but that would go beyond this PR so it won't affect this, though I'm happy to hear if you have any feelings after having tooled around here.

workingjubilee commented on pgrx-tests/tests/ui/invalid_pgcast_options.rs:
```diff
@@ -0,0 +1,8 @@
+use pgrx::prelude::*;
+
+#[pg_cast(invalid_opt)]
+pub fn cast_function(foo: i32) -> i32 {
+    foo
+}
+
+fn main() {}
```
This one fails to compile (which is correct), please run:
```shell
TRYBUILD=overwrite cargo test --test ui
```
and then commit the resulting .stderr file.

xwkuang5 replied:
Done

workingjubilee commented on pgrx-tests/tests/ui/invalid_pgcast_function.rs:
```diff
@@ -0,0 +1,8 @@
+#[pg_cast]
+pub fn cast_function() -> i32 {
+    0
+}
```
This one successfully compiles (which is incorrect). The implementation needs to validate the argument count, it appears. When you have made this fail to compile, please run the same command as I described for the other test.

xwkuang5 replied:
Removed this one because this is a runtime error that only get triggered when we actually generate some SQL entities. Therefore it does not fit as a ui test.

workingjubilee replied:
Oh okay! I sometimes don't remember what actually happens during compilation and what is technically """runtime""". I'd like to add schema-gen testing to the UI work, but I think that may need a burlier library than trybuild to do it, so it's definitely a separate issue. Thanks!

workingjubilee commented on pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:
```diff
@@ -472,6 +474,119 @@ impl ToSql for PgExternEntity {
+                    "PG cast function must have exactly one argument, got {}",
+                    self.fn_args.len()
+                ));
+            }
```
Somehow, this does nothing. I'm not sure what's going on there, honestly. My "essentially just instinct" guess is that this is somehow skipped over outright during building the example function used, and the closest reason why seems to be that the `else if let Some` was missed?

Is it correct for it to be an `else if let`?

xwkuang5 replied:
Updated to `if let`. FWIW, this does do what we expect (output below), it's just this is only triggered at SQL generation time (see the other comment in the UI test). 

```
#[pg_cast]
pub fn cast_function() -> bool {
    true
}

    Finished dev [unoptimized + debuginfo] target(s) in 3.28s
 Discovering SQL entities
  Discovered 2 SQL entities: 0 schemas (0 unique), 2 functions, 0 types, 0 enums, 0 sqls, 0 ords, 0 hashes, 0 aggregates, 0 triggers
     Writing SQL entities to /dev/stdout
Error:
   0: Could not write SQL to stdout
   1: PG cast function (cast_function) must have exactly one argument, got 0

Location:
   pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:515
```

workingjubilee reviewed (changes requested):
Apologies, but this does seem to be unfinished! Thank you for writing the tests! I think it's probably just a missed condition check.

xwkuang5 committed:
Add ui test output for `invalid_pgcast_options` test.

xwkuang5 committed:
Remove ui test `invalid_pgcast_function`.

The validation against the entity itself is done at runtime (during SQL
generation). Hence it will not be triggered at compile time. Here's the
output from a local test with an extension:

```
$/home/louiskuang/Documents/pgrx/target/debug/cargo-pgrx pgrx schema --features "pg13"
       Using FeatureFlag("pg13") and `pg_config` from /home/louiskuang/.pgrx/13.13/pgrx-install/bin/pg_config
    Building for SQL generation with features `pg13`
   Compiling pgrx-sql-entity-graph v0.11.2 (/home/louiskuang/Documents/pgrx/pgrx-sql-entity-graph)
   Compiling pgrx-macros v0.11.2 (/home/louiskuang/Documents/pgrx/pgrx-macros)
   Compiling pgrx-pg-sys v0.11.2 (/home/louiskuang/Documents/pgrx/pgrx-pg-sys)
   Compiling pgrx v0.11.2 (/home/louiskuang/Documents/pgrx/pgrx)
   Compiling my_extension v0.0.0 (/home/louiskuang/Documents/my_extension)
warning: unused import: `Json`
 --> src/lib.rs:1:24
  |
1 | use pgrx::{prelude::*, Json};
  |                        ^^^^
  |
  = note: `#[warn(unused_imports)]` on by default

warning: `my_extension` (lib) generated 1 warning (run `cargo fix --lib -p my_extension` to apply 1 suggestion)
    Finished dev [unoptimized + debuginfo] target(s) in 11.21s
 Discovering SQL entities
  Discovered 2 SQL entities: 0 schemas (0 unique), 2 functions, 0 types, 0 enums, 0 sqls, 0 ords, 0 hashes, 0 aggregates, 0 triggers
     Writing SQL entities to /dev/stdout
Error:
   0: Could not write SQL to stdout
   1: PG cast function (cast_function) must have exactly one argument, got 0

Location:
   pgrx-sql-entity-graph/src/pg_extern/entity/mod.rs:515
```

xwkuang5 committed:
Make SQL generation for `pg_cast` independent of `pg_operator`.

While `pg_cast` and `pg_operator` are indeed mutually-exclusive currently (
`pg_operator` always requires a function with two parameters and `pg_cast`
always requires a function with exactly one parameter), there's no
reason why they can not both apply on the same element, say in the
future the need for non-binary operator arises.

In addition, both `pg_cast` and `pg_operator` are doing its own
validation at the moment so there's no risk of being in an impossible
state like the following:

```
pub fn foo(left: i32, right: i32) -> bool {
    true
}
```

workingjubilee reviewed (approved):
Great!

xwkuang5 commented:
Thank you for the review!
